File: blasd/syr2.c

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

#include <stdio.h>
#include <stdint.h>

#include "dtype.h"
#include "interfaces.h"
#include "mvec_nosimd.h"


void __update_syr2_recursive(mdata_t *A, const mvec_t *X, const mvec_t *Y,
                             DTYPE alpha, int flags, int N)
{
  mvec_t x0, y0;
  mdata_t A0;

  if (N < MIN_MVEC_SIZE) {
    __update_trmv_unb(A, X, Y, alpha, flags, N, N);
    __update_trmv_unb(A, Y, X, alpha, flags, N, N);
    return;
  }

  __subvector(&x0, X, 0);
  __subvector(&y0, Y, 0);
  __subblock(&A0, A, 0, 0);
  if (N/2 < MIN_MVEC_SIZE) {
    __update_trmv_unb(&A0, &x0, &y0, alpha, flags, N/2, N/2);
    __update_trmv_unb(&A0, &y0, &x0, alpha, flags, N/2, N/2);
  } else {
    __update_syr2_recursive(&A0, &x0, &y0, alpha, flags, N/2);
  }

  if (flags & GOMAS_UPPER) {
    __subvector(&x0, X, 0);
    __subvector(&y0, Y, N/2);
    __subblock(&A0, A, 0, N/2);
    __update_ger_recursive(&A0, &x0, &y0, alpha, flags, N-N/2, N/2);
    __subvector(&x0, X, N/2);
    __subvector(&y0, Y, 0);
    __update_ger_recursive(&A0, &y0, &x0, alpha, flags, N-N/2, N/2);
  } else {
    __subvector(&y0, Y, 0);
    __subvector(&x0, X, N/2);
    __subblock(&A0, A, N/2, 0);
    __update_ger_recursive(&A0, &x0, &y0, alpha, flags, N/2, N-N/2);
    __subvector(&y0, Y, N/2);
    __subvector(&x0, X, 0);
    __update_ger_recursive(&A0, &y0, &x0, alpha, flags, N/2, N-N/2);
  }

  __subvector(&y0, Y, N/2);
  __subvector(&x0, X, N/2);
  __subblock(&A0, A, N/2, N/2);
  if (N-N/2 < MIN_MVEC_SIZE) {
    __update_trmv_unb(&A0, &x0, &y0, alpha, flags, N-N/2, N-N/2);
    __update_trmv_unb(&A0, &y0, &x0, alpha, flags, N-N/2, N-N/2);
  } else {
    __update_syr2_recursive(&A0, &x0, &y0, alpha, flags, N-N/2);
  }
}


// Local Variables:
// indent-tabs-mode: nil
// End:

File: blasd/syr2.go

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

package blasd

// #cgo CFLAGS: -O3 -march=native -fomit-frame-pointer -ffast-math -Iinc -I.
// #cgo LDFLAGS: -lm
// #include "inc/interfaces.h"
import  "C"
import "unsafe"
import "github.com/hrautila/cmat"
import "github.com/hrautila/gomas"

func syr2(A, X, Y *cmat.FloatMatrix, alpha float64, bits, N int) error {
    var Am C.mdata_t
    var Xm, Ym C.mvec_t
    
    xr, _ := X.Size()
    yr, _ := Y.Size()
    Am.md = (*C.double)(unsafe.Pointer(&A.Data()[0]))
    Am.step = C.int(A.Stride())
    Xm.md = (*C.double)(unsafe.Pointer(&X.Data()[0]))
    Ym.md = (*C.double)(unsafe.Pointer(&Y.Data()[0]))
    Ym.inc = C.int(1)
    Xm.inc = C.int(1)

    // if row vectors, change increment
    if xr == 1 {
        Xm.inc = C.int(X.Stride())
    }
    if yr == 1 {
        Ym.inc = C.int(Y.Stride())
    }

    C.__d_update_syr2_recursive(
        (*C.mdata_t)(unsafe.Pointer(&Am)), 
        (*C.mvec_t)(unsafe.Pointer(&Xm)),
        (*C.mvec_t)(unsafe.Pointer(&Ym)), 
        C.double(alpha), C.int(bits), C.int(N))
    return nil
}


func MVUpdate2Sym(A, X, Y *cmat.FloatMatrix, alpha float64, bits int, confs... *gomas.Config) *gomas.Error {
    ar, ac := A.Size()
    yr, yc := Y.Size()
    xr, xc := X.Size()

    if ar*ac == 0 {
        return nil
    }
    if yr != 1 && yc != 1 {
        return gomas.NewError(gomas.ENEED_VECTOR, "MVUpdate2Sym")
    }
    if xr != 1 && xc != 1 {
        return gomas.NewError(gomas.ENEED_VECTOR, "MVUpdate2Sym")
    }
    nx := X.Len()
    ny := Y.Len()
    if ac != ar || ac != ny || ny != nx {
        return gomas.NewError(gomas.ESIZE, "MVUpdate2Sym")
    }
    syr2(A, X, Y, alpha, bits, ny)
    return nil
}

// Local Variables:
// tab-width: 4
// indent-tabs-mode: nil
// End: